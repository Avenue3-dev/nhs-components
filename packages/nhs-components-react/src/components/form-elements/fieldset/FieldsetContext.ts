import { createContext, useContext } from 'react'

export type FieldsetContextValue = {
  isFieldset: boolean
  setErrored: (componentId: string) => void
  resetErrored: (componentId: string) => void
  register: (componentId: string) => void
  unregister: (componentId: string) => void
}

const FieldsetContext = createContext<FieldsetContextValue>({
  isFieldset: false,
  setErrored: () => {},
  resetErrored: () => {},
  register: () => {},
  unregister: () => {},
})

export const useFieldsetContext = () =>
  useContext<FieldsetContextValue>(FieldsetContext)

export default FieldsetContext
