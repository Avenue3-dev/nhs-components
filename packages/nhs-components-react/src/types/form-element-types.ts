import { ErrorMessageProps } from '@/components/form-elements/error-message/ErrorMessage'
import { LabelProps } from '@/components/miscellaneous/label/Label'
import { HTMLProps, ReactNode } from 'react'

export type FormElementProps = {
  label?: string
  labelProps?: LabelProps
  error?: string | boolean
  errorProps?: ErrorMessageProps
  hint?: ReactNode
  hintProps?: HTMLProps<HTMLDivElement>
  formGroupProps?: HTMLProps<HTMLDivElement>
  disableErrorLine?: boolean
  id?: string
  name?: string
}
