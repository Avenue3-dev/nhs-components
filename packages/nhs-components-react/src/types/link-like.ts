import { ElementType, HTMLProps } from 'react'

export type AsElementLink<T extends HTMLAnchorElement> = HTMLProps<T> & {
  asElement?: ElementType | ElementType<T>
  to?: unknown
}
