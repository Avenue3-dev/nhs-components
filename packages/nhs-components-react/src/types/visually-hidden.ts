export type VisuallyHiddenProps = {
  visuallyHiddenText: string;
};
