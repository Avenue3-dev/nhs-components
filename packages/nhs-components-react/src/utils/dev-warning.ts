const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
