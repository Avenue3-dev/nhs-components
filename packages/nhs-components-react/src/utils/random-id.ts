export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
