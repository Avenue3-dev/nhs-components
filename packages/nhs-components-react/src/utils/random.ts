export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
