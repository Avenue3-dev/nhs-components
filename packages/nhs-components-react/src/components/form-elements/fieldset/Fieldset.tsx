import clsx from 'clsx'
import { ElementType, HTMLProps, useState } from 'react'
import FieldsetContext, { FieldsetContextValue } from './FieldsetContext'
import { Size } from '@/types/nhsuk-sizes'

type LegendProps = {
  isPageHeading?: boolean
  asElement?: ElementType
  size?: Size
} & Omit<HTMLProps<HTMLLegendElement>, 'size'>

type Legend = React.FC<LegendProps>

const Legend: Legend = ({
  className,
  children,
  isPageHeading,
  asElement: Component = 'h1',
  size,
  ...rest
}): JSX.Element => {
  return (
    <legend
      className={clsx(
        'nhsuk-legend',
        {
          'nhsuk-fieldset__legend--xl': isPageHeading && !size,
        },
        { [`nhsuk-fieldset__legend--${size}`]: size },
        className,
      )}
      {...rest}
    >
      {isPageHeading ? (
        <Component className="nhsuk-fieldset__heading">{children}</Component>
      ) : (
        children
      )}
    </legend>
  )
}

type FieldsetProps = {
  disableErrorLine?: boolean
} & HTMLProps<HTMLFieldSetElement>

type Fieldset = {
  Legend: Legend
} & React.FC<FieldsetProps>

/**
 * Use a fieldset to group related form inputs.
 *
 * See more information regarding this component [here](https://service-manual.nhs.uk/design-system/components/fieldset).
 */
const Fieldset: Fieldset = ({
  className,
  children,
  disableErrorLine,
  ...rest
}): JSX.Element => {
  const [registered, setRegistered] = useState<string[]>([])
  const [errored, setErrored] = useState<string[]>([])

  const erroredComponent = (componentId: string, add: boolean) => {
    const hasError = errored.includes(componentId)

    if (add && !hasError) {
      setErrored([...errored, componentId])
    }

    if (!add && hasError) {
      setErrored(errored.filter((id) => id !== componentId))
    }
  }

  const registerComponent = (componentId: string, add: boolean) => {
    const haveRegistered = registered.includes(componentId)
    if (add && !haveRegistered) {
      setRegistered([...registered, componentId])
    }

    if (!add && haveRegistered) {
      setRegistered(registered.filter((id) => id !== componentId))
    }
  }

  const contextValue: FieldsetContextValue = {
    isFieldset: true,
    setErrored: (id) => erroredComponent(id, true),
    resetErrored: (id) => erroredComponent(id, false),
    register: (id) => registerComponent(id, true),
    unregister: (id) => registerComponent(id, false),
  }

  const BaseFieldset = (): JSX.Element => {
    return (
      <fieldset className={clsx('nhsuk-fieldset', className)} {...rest}>
        {children}
      </fieldset>
    )
  }

  return (
    <FieldsetContext.Provider value={contextValue}>
      {registered.length > 0 ? (
        <div
          className={clsx('nhsuk-form-group', {
            'nhsuk-form-group--error': disableErrorLine
              ? false
              : errored.length > 0,
          })}
        >
          <BaseFieldset />
        </div>
      ) : (
        <BaseFieldset />
      )}
    </FieldsetContext.Provider>
  )
}

Fieldset.Legend = Legend

export default Fieldset
