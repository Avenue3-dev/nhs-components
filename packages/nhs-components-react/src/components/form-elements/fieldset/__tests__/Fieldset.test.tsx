import { render, screen } from '@/utils/test-utils'
import { describe, test, expect } from 'vitest'
import Fieldset from '../Fieldset'
import { Size } from '@/types/nhsuk-sizes'
import { useFieldsetContext } from '../FieldsetContext'
import { useEffect } from 'react'

describe('Fieldset', () => {
  test('should match the snapshot', () => {
    render(<Fieldset data-testid="fieldset" />)

    expect(screen.getByTestId('fieldset')).toMatchSnapshot()
  })

  test('should match the snapshot - have registered components', () => {
    const Sample = () => {
      const { register } = useFieldsetContext()

      useEffect(() => {
        register('component')
      }, [])

      return 'sample'
    }

    render(<Fieldset data-testid="fieldset">
      <Sample />
    </Fieldset>)

    expect(screen.getByTestId('fieldset').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - errored', () => {
    const Sample = () => {
      const { register, setErrored } = useFieldsetContext()

      useEffect(() => {
        register('component')
        setErrored('component')
      }, [])

      return 'sample'
    }

    render(<Fieldset data-testid="fieldset">
      <Sample />
    </Fieldset>)

    expect(screen.getByTestId('fieldset').parentNode).toMatchSnapshot()
  })
})

describe('Fieldset.Legend', () => {
  test('should match the snapshot', () => {
    render(
      <Fieldset.Legend data-testid="fieldset-legend">Legend</Fieldset.Legend>,
    )

    expect(screen.getByTestId('fieldset-legend')).toMatchSnapshot()
  })

  test('should match the snapshot - isPageHeading', () => {
    render(
      <Fieldset.Legend data-testid="fieldset-legend" isPageHeading>
        Legend
      </Fieldset.Legend>,
    )

    expect(screen.getByTestId('fieldset-legend')).toMatchSnapshot()
  })

  test('should match the snapshot - isPageHeading as h2', () => {
    render(
      <Fieldset.Legend
        data-testid="fieldset-legend"
        isPageHeading
        asElement="h2"
      >
        Legend
      </Fieldset.Legend>,
    )

    expect(screen.getByTestId('fieldset-legend')).toMatchSnapshot()
  })

  test.each(['s', 'm', 'l', 'xl'] as Size[])(
    'should match the snapshot - size %s',
    (size: Size) => {
      render(
        <Fieldset.Legend data-testid="fieldset-legend" size={size}>
          Legend
        </Fieldset.Legend>,
      )

      expect(screen.getByTestId('fieldset-legend')).toMatchSnapshot()
    },
  )
})
