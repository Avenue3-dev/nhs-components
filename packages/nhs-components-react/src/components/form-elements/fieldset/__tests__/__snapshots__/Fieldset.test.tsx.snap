// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fieldset > should match the snapshot - errored 1`] = `
<div
  class="nhsuk-form-group nhsuk-form-group--error"
>
  <fieldset
    class="nhsuk-fieldset"
    data-testid="fieldset"
  >
    sample
  </fieldset>
</div>
`;

exports[`Fieldset > should match the snapshot - have registered components 1`] = `
<div
  class="nhsuk-form-group"
>
  <fieldset
    class="nhsuk-fieldset"
    data-testid="fieldset"
  >
    sample
  </fieldset>
</div>
`;

exports[`Fieldset > should match the snapshot 1`] = `
<fieldset
  class="nhsuk-fieldset"
  data-testid="fieldset"
/>
`;

exports[`Fieldset.Legend > should match the snapshot - isPageHeading 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--xl"
  data-testid="fieldset-legend"
>
  <h1
    class="nhsuk-fieldset__heading"
  >
    Legend
  </h1>
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot - isPageHeading as h2 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--xl"
  data-testid="fieldset-legend"
>
  <h2
    class="nhsuk-fieldset__heading"
  >
    Legend
  </h2>
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot - size l 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--l"
  data-testid="fieldset-legend"
>
  Legend
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot - size m 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--m"
  data-testid="fieldset-legend"
>
  Legend
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot - size s 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--s"
  data-testid="fieldset-legend"
>
  Legend
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot - size xl 1`] = `
<legend
  class="nhsuk-legend nhsuk-fieldset__legend--xl"
  data-testid="fieldset-legend"
>
  Legend
</legend>
`;

exports[`Fieldset.Legend > should match the snapshot 1`] = `
<legend
  class="nhsuk-legend"
  data-testid="fieldset-legend"
>
  Legend
</legend>
`;
